tests/e2e/contracts/delegation/delegation.sol

```diff
@@ -0,0 +1,252 @@
+            } else if (epochLow == 24) {
+                epoch = uint64(uint8(result[fieldOffset + 1]));
+                fieldOffset += 2;
+                require(epoch >= 24, "malformed epoch in receipt");
+            } else {
+                // A real implementation would support decoding bigger epoch numbers.
+                revert("unsupported epoch length");
+            }
+
+            // Decode end receipt identifier.
+            require(result[fieldOffset] == 0x67 && result[fieldOffset + 1] == "r", "malformed receipt");
+            uint8 endReceipt = uint8(result[fieldOffset + 8]) & 0x1f; // Assume end receipt is never greater than 23.
+
+            pendingUndelegations[receiptId].endReceiptId = endReceipt;
+            undelegationPools[endReceipt].totalShares += pending.shares;
+        } else {
+            // Undelegation failed to start, return the shares.
+            delegations[msg.sender][pending.from] += pending.shares;
+            delete(pendingUndelegations[receiptId]);
+        }
+    }
+
+    function undelegateDone(uint8 receiptId) public {
+        require(receiptId <= 23, "receipt identifier overflow"); // Because our CBOR encoder is lazy.
+
+        PendingUndelegation memory pending = pendingUndelegations[receiptId];
+        require(pending.to != address(0), "unknown receipt");
+        require(pending.endReceiptId > 0, "must call undelegateStart first");
+
+        UndelegationPool memory pool = undelegationPools[pending.endReceiptId];
+        if (pool.totalAmount == 0) {
+            // Did not fetch the end receipt yet, do it now.
+            bytes memory result = takeReceipt(RECEIPT_KIND_UNDELEGATE_DONE, pending.endReceiptId);
+
+            // This is a very lazy CBOR decoder. It assumes that if there is only an amount field then
+            // the undelegation operation succeeded and if not, there was some sort of error which is
+            // not decoded.
+            uint128 amount = 0;
+            if (result[0] == 0xA1 && result[1] == 0x66 && result[2] == "a") {
+                // Undelegation succeeded, decode token amount.
+                uint8 amountLen = uint8(result[8]) & 0x1f; // Assume amount field is never greater than 16 bytes.
+                for (uint offset = 0; offset < amountLen; offset++) {
+                    uint8 v = uint8(result[9 + offset]);
+                    amount += uint128(v) << 8*uint128(amountLen - offset - 1);
+                }
+
+                undelegationPools[pending.endReceiptId].totalAmount = amount;
+                pool.totalAmount = amount;
+            } else {
+                // Should never fail.
+                revert("undelegation failed");
+            }
+        }
+
+        // Compute how much we get from the pool and transfer the amount.
+        uint128 transferAmount = (pending.shares * pool.totalAmount) / pool.totalShares;
+        pending.to.transfer(transferAmount);
+
+        delete(pendingUndelegations[receiptId]);
+    }
+}
```

Comment:
strictly speaking, for just testing, I would have passed the cbor directly into the contract and have it proxy the subcall instead. would have been a bit clearer what's going on, too.

Reply:
Yeah thought about that, I just wanted to write some Solidity to see if the API was viable. Any thoughts about the API, does it make sense?

Reply:
Could use an existing solidity cbor lib? That would make it clearer to show how it would look like in practice. Unless there's too much work incorporating that into the existing unit tests.

Reply:
I'm in the process of re-working the CBOR stuff (it was an easy fix to support 64bit ints, but need to do better parsing).

I'm happy with the unit test as a starting point and will be submitting a PR for sapphire-paratime repo when the updated sapphire-dev image is available.

Reply:
> Could use an existing solidity cbor lib? That would make it clearer to show how it would look like in practice. Unless there's too much work incorporating that into the existing unit tests.

Not easily for these E2E tests, we would need to adapt the pipeline a bit.